File: src/main/java/Browser.java
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.Point;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.FluentWait;
import org.openqa.selenium.support.ui.Wait;
import org.openqa.selenium.support.ui.WebDriverWait;

public class Browser 
{
	private String _mainWindowHandler;
	private WebDriver Driver;
	private String Type;
	public Browser() 
	{
		Type = "Chrome";
	}
	
	public Browser(String type)
	{
		Type = type;
	}

	public void StartWebDriver() 
	{
		switch(Type)
		{
			case "Chrome":
			{
				Driver = StartChrome();
				break;
			}
			case "Firefox":
			{
				Driver = StartFirefox();
				break;
			}
		}
		Driver.manage().deleteAllCookies();
		Driver.manage().timeouts().setScriptTimeout(60, TimeUnit.SECONDS);
		Driver.manage().timeouts().pageLoadTimeout(60, TimeUnit.SECONDS);
		Driver.manage().window().maximize();
		_mainWindowHandler = Driver.getWindowHandle();
	}

	private FirefoxDriver StartFirefox()
	{
		Path currentRelativePath = Paths.get("");
		String s = currentRelativePath.toAbsolutePath().toString();
		
		System.setProperty("webdriver.gecko.driver",s+"\\geckodriver.exe");
		return new FirefoxDriver();
	}
	
	private ChromeDriver StartChrome() 
	{
		Path currentRelativePath = Paths.get("");
		String s = currentRelativePath.toAbsolutePath().toString();

		System.setProperty("webdriver.chrome.driver",s+"\\chromedriver.exe");
		ChromeOptions chromeOptions = new ChromeOptions();
		chromeOptions.addArguments("download.default_directory", "C:\\TestData");
		chromeOptions.addArguments("safebrowsing.enabled", "false");
		chromeOptions.addArguments("acceptInsecureCerts", "true");
		chromeOptions.addArguments("disable-extensions");
		chromeOptions.addArguments("ignore-certificate-errors", "ignore-urlfetcher-cert-requests",
				"allow-insecure-localhost");
		return new ChromeDriver(chromeOptions);
	}

	public WebElement FindElement(String XPath) 
	{
		WaitForAjaxComplete(4);
		WebElement element;

		Wait<WebDriver> wait = new FluentWait<WebDriver>(Driver)
				.withTimeout(60, TimeUnit.SECONDS)
				.pollingEvery(500, TimeUnit.MILLISECONDS)
				.ignoring(Exception.class);
		try 
		{
			element = wait.until(ExpectedConditions.elementToBeClickable(By.xpath(XPath)));
		} 
		catch (Exception e) 
		{
			//здесь был логгер, но его нет желани€ доделывать (так же везде должно быть)
			return null;
		}

		return element;
	}

	public ArrayList<WebElement> FindElements(String XPath) 
	{
		WaitForAjaxComplete(15);
		ArrayList<WebElement> elements = new ArrayList<WebElement>();

		Wait<WebDriver> wait = new FluentWait<WebDriver>(Driver)
				.withTimeout(60, TimeUnit.SECONDS)
				.pollingEvery(1, TimeUnit.MILLISECONDS)
				.ignoring(Exception.class);

		try 
		{
			List<WebElement> finded = wait.until(ExpectedConditions.visibilityOfAllElementsLocatedBy(By.xpath(XPath)));
			for (WebElement elem : finded) 
			{
				elements.add(elem);
			}
		} 
		catch (Exception e)
		{
			//здесь был логгер, но его нет желани€ доделывать (так же везде должно быть)
			return null;
		}

		return elements;
	}

	public WebElement FindHiddenElement(String XPath, int timeOut) 
	{
		WaitForAjaxComplete(5);
		WebElement element;

		Wait<WebDriver> wait = new FluentWait<WebDriver>(Driver)
				.withTimeout(60, TimeUnit.SECONDS)
				.pollingEvery(1, TimeUnit.MILLISECONDS)
				.ignoring(Exception.class);
		try 
		{
			element = wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath(XPath)));
		} 
		catch (Exception e) 
		{
			//здесь был логгер, но его нет желани€ доделывать (так же везде должно быть)
			return null;
		}

		return element;
	}

	public void Navigate(String Url) 
	{
		Driver.navigate().to(Url);
	}

	public void SwitchToMain()
	{
		Driver.switchTo().window(_mainWindowHandler);
	}

	public void SwitchToNewWindow()
	{
		Set<String> handles = Driver.getWindowHandles();
		List<String> list = new ArrayList<String>(handles);
		String handle = list.get(list.size() - 1);
		Driver.switchTo().window(handle);
	}

	public Set<String> GetHandlers()
	{
		return Driver.getWindowHandles();
	}

	public void SwitchToNewWindow(String Handl)
	{
		Driver.switchTo().window(Handl);
	}

	public void CloseActiveWindows() 
	{
		Set<String> handles = Driver.getWindowHandles();
		for (String handl : handles) 
		{
			handles = Driver.getWindowHandles();
			SwitchToNewWindow(handl);
			if (handl != _mainWindowHandler) 
			{
				Driver.close();
			}
		}
	}

	public void OpenWindow() 
	{
		ExecuteJavaScript("window.open('your url','_blank');", null);
	}

	public void Refresh() 
	{
		Driver.navigate().refresh();
	}

	public String getUrl() 
	{
		WaitForAjaxComplete(3);
		return Driver.getCurrentUrl();
	}

	public void Quit() 
	{
		Driver.quit();
	}

	public void AllertAccept() 
	{
		WebDriverWait wait = new WebDriverWait(Driver, 30 /* timeout in seconds */);
		if (wait.until(ExpectedConditions.alertIsPresent()) != null) 
		{
			Driver.switchTo().alert().accept();
		}
	}

	public Object ExecuteJavaScript(String javaScript, Object element) 
	{
		JavascriptExecutor javaScriptExecutor = (JavascriptExecutor) Driver;
		try
		{
			return javaScriptExecutor.executeScript(javaScript, element);
		} 
		catch (Exception e)
		{
			return javaScriptExecutor.executeScript(javaScript, element);
		}
	}

	public void WaitForAjaxComplete(int timeoutInSeconds)
	{
		//метод изменен. —тарый в €ве не заработал.
		String firstSource = Driver.getPageSource();
		try
		{
			Thread.sleep(100);
		}
		catch(Exception e)
		{
			//nothing to do now
		}
		String secondSource = Driver.getPageSource();
		int allTime = 0;
		while(firstSource != secondSource)
		{
			if(allTime >= timeoutInSeconds)
			{
				break;
			}
			firstSource = Driver.getPageSource();
			try
			{
				Thread.sleep(100);
			}
			catch(Exception ex)
			{
				//nothing to do now
			}
			secondSource = Driver.getPageSource();
			allTime+=100;
		}
	}	

	public void ScrollElementToCenter(Point location)
	{
		int height = GetBrowserWindowInnerHeight() / 2;
		location.y = location.y - height;
		ScrollToElement(location);
	}

	public void ScrollToElement(Point location) 
	{
		ExecuteJavaScript("window.scrollTo("+location.x+","+ location.y+");", null);
	}

	public int GetBrowserWindowInnerHeight() 
	{
		return  Integer.parseInt(ExecuteJavaScript("return window.innerHeight;", null).toString());
	}
}

File: Autotests/Browser/Browser.cs
using OpenQA.Selenium;
using System;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Support.UI;
using System.Collections.Generic;
using System.Drawing;
using System.Threading;
using System.Linq;
using System.Collections.ObjectModel;
using OpenQA.Selenium.Remote;

namespace Core
{
    public class Browser
    {
        private string _mainWindowHandler { get; set; }
        private IWebDriver Driver { get; set; }
        public Browser()
        {

        }

        public void StartWebDriver()
        {
            Driver = StartChrome();

            Driver.Manage().Cookies.DeleteAllCookies();
            Driver.Manage().Timeouts().AsynchronousJavaScript = TimeSpan.FromSeconds(60);
            Driver.Manage().Timeouts().PageLoad = TimeSpan.FromSeconds(60);
            Driver.Manage().Window.Maximize();
            _mainWindowHandler = Driver.CurrentWindowHandle;
        }

        private ChromeDriver StartChrome()
        {
            string dir = System.IO.Directory.GetCurrentDirectory();
            ChromeOptions chromeOptions = new ChromeOptions();
            chromeOptions.AddUserProfilePreference("download.default_directory", @"C:\TestData");
            chromeOptions.AddUserProfilePreference("safebrowsing.enabled", "false");
            chromeOptions.AddUserProfilePreference("acceptInsecureCerts", "true");
            chromeOptions.AddArguments("disable-extensions");
            chromeOptions.AddArguments("ignore-certificate-errors", "ignore-urlfetcher-cert-requests", "allow-insecure-localhost");
            return new ChromeDriver(dir, chromeOptions);
        }

        public IWebElement FindElement(string XPath)
        {
            WaitForAjaxComplete(5);
            IWebElement element;

            DefaultWait<IWebDriver> wait = new DefaultWait<IWebDriver>(Driver)
            {
                Timeout = TimeSpan.FromSeconds(60),
                PollingInterval = TimeSpan.FromMilliseconds(1)
            };
            try
            {
                wait.IgnoreExceptionTypes(typeof(Exception));
                element = wait.Until(condition: SeleniumExtras.WaitHelpers.ExpectedConditions.ElementToBeClickable(By.XPath(XPath)));
            }
            catch
            {
                return null;
            }

            return element;
        }

        public List<IWebElement> FindElements(string XPath)
        {
            WaitForAjaxComplete(5);
            List<IWebElement> elements = new List<IWebElement>();

            DefaultWait<IWebDriver> wait = new DefaultWait<IWebDriver>(Driver)
            {
                Timeout = TimeSpan.FromSeconds(60),
                PollingInterval = TimeSpan.FromMilliseconds(1)
            };
            wait.IgnoreExceptionTypes(typeof(Exception));
            try
            {
                ReadOnlyCollection<IWebElement> finded = wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.VisibilityOfAllElementsLocatedBy(By.XPath(XPath)));
                foreach (IWebElement elem in finded)
                {
                    elements.Add(elem);
                }
            }
            catch
            {
                return null;
            }

            return elements;
        }

        public List<IWebElement> FindHiddenElements(string XPath)
        {
            WaitForAjaxComplete(5);
            List<IWebElement> elements = new List<IWebElement>();

            DefaultWait<IWebDriver> wait = new DefaultWait<IWebDriver>(Driver)
            {
                Timeout = TimeSpan.FromSeconds(60),
                PollingInterval = TimeSpan.FromMilliseconds(1)
            };
            wait.IgnoreExceptionTypes(typeof(Exception));
            Func<IWebDriver, ReadOnlyCollection<IWebElement>> finder = new Func<IWebDriver, ReadOnlyCollection<IWebElement>>(x => x.FindElements(By.XPath(XPath)));
            try
            {
                ReadOnlyCollection<IWebElement> finded = wait.Until(finder);
                foreach (IWebElement elem in finded)
                {
                    elements.Add(elem);
                }
            }
            catch
            {
                return null;
            }

            return elements;
        }
        private List<IWebElement> CheckConditions(string XPath)
        {
            ReadOnlyCollection<IWebElement> elements = Driver.FindElements(By.XPath(XPath));

            List<IWebElement> result = new List<IWebElement>();

            foreach (IWebElement elem in elements)
            {
                if (elem.Enabled && elem.Displayed)
                {
                    result.Add(elem);
                }
            }

            return result;
        }


        public IWebElement FindHiddenElement(string XPath, int timeOut = 15)
        {
            WaitForAjaxComplete(5);
            IWebElement element;

            DefaultWait<IWebDriver> wait = new DefaultWait<IWebDriver>(Driver)
            {
                Timeout = TimeSpan.FromSeconds(timeOut),
                PollingInterval = TimeSpan.FromMilliseconds(1)
            };
            wait.IgnoreExceptionTypes(typeof(Exception));
            try
            {
                element = wait.Until(condition: SeleniumExtras.WaitHelpers.ExpectedConditions.ElementExists(By.XPath(XPath)));
            }
            catch
            {
                return null;
            }

            return element;
        }

        public void Navigate(string Url)
        {
            UriBuilder uri = new UriBuilder(Url);
            Driver.Navigate().GoToUrl(uri.Uri);
        }

        public void SwitchToMain()
        {
            Driver.SwitchTo().Window(_mainWindowHandler);
        }
        public void SwitchToNewWindow()
        {
            ReadOnlyCollection<string> handles = Driver.WindowHandles;
            Driver.SwitchTo().Window(handles.Last());
        }

        public IReadOnlyCollection<string> GetHandlers()
        {
            return Driver.WindowHandles;
        }

        public void SwitchToNewWindow(string Handl)
        {
            Driver.SwitchTo().Window(Handl);
        }

        public void CloseActiveWindows()
        {
            ReadOnlyCollection<string> handles = Driver.WindowHandles;
            foreach (string handl in handles)
            {
                handles = Driver.WindowHandles;
                SwitchToNewWindow(handl);
                if (handl != _mainWindowHandler)
                {
                    Driver.Close();
                }
            }
        }

        public void OpenWindow()
        {
            ExecuteJavaScript("window.open('your url','_blank');");
        }

        public void Refresh()
        {
            Driver.Navigate().Refresh();
        }

        public Uri Url
        {
            get
            {
                WaitForAjaxComplete(3);
                return new Uri(Driver.Url);
            }
        }

        public string Title
        {
            get
            {
                WaitForAjaxComplete(3);
                return Driver.Title;
            }
        }
        public void Quit()
        {
            Driver.Quit();
        }

        public void AllertAccept()
        {
            WebDriverWait wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(30) /*timeout in seconds*/);
            if (wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.AlertIsPresent()) != null)
            {
                Driver.SwitchTo().Alert().Accept();
            }
        }

        public object ExecuteJavaScript(string javaScript, params object[] args)
        {
            IJavaScriptExecutor javaScriptExecutor = (IJavaScriptExecutor)Driver;
            try
            {
                return javaScriptExecutor.ExecuteScript(javaScript, args);
            }
            catch
            {
                return javaScriptExecutor.ExecuteScript(javaScript, args);
            }
        }
        public void WaitForAjaxComplete(int timeoutInSeconds)
        {
            string firstSource = Driver.PageSource;
            Thread.Sleep(100);
            string secondSource = Driver.PageSource;
            int allTime = 0;
            while(firstSource != secondSource)
            {
                if(allTime >= timeoutInSeconds)
                {
                    break;
                }
                firstSource = Driver.PageSource;
                Thread.Sleep(100);
                secondSource = Driver.PageSource;
                allTime += 100;
            }
        }
        public void ScrollElementToCenter(Point location)
        {
            ScrollToElement(location - new Size(0, GetBrowserWindowInnerHeight() / 2));
        }
        public void ScrollToElement(Point location)
        {
            ExecuteJavaScript($"window.scrollTo({location.X}, {location.Y});");
        }
        public int GetBrowserWindowInnerHeight()
        {
            return int.Parse(ExecuteJavaScript("return window.innerHeight;").ToString());
        }
    }
}
